fix: empty slice lists of record arrays (#1593)

agoose77 opened the pull request:
This PR fixes #1593, and closes #1601, in the following ways:
- Normalises the `[]` index into an integer array, so that it passes through the correct indexing logic
- Ensures the zero-length is set for advanced slices
- Prevents [NumPy advanced-basic mixed indexing](https://docs.scipy.org/doc/numpy/reference/arrays.indexing.html#combining-advanced-and-basic-indexing) (case 1)
- Hardens the `_getitem` logic against non-sized iterables.
   - Normally this is best to let fail with the `len()` check, but in this case we are effectively implementing a `match` and it's much more elegant to ask for permission.

The existing test suite has been modified with this PR. Only one test has a modified expected output:
https://github.com/scikit-hep/awkward/blob/456bae3de7d0edf82c34a2e948b76a7a786cd2ac/tests/v2/test_0024-use-regular-array.py#L23

Some tests now raise intentional errors, so these are required using `pytest.raises`.

agoose77 committed:
Fix: tentative fix for empty slice lists

agoose77 committed:
Test: validate fix

ioanaif commented on src/awkward/_v2/contents/recordarray.py:
```diff
@@ -315,7 +315,11 @@ def _getitem_field(self, where, only_fields=()):
+        if len(where) == 0:
+            indexes = range(len(self._contents))
+        else:
+            indexes = [self.field_to_index(field) for field in where]
```
```suggestion
            indexes = [self.field_to_index(field) for field in where]
            if self._fields is None:
                fields = None
            else:
                if len(where) == 0:
                    fields = [self._fields[i] for i in range(len(self._contents))]
                else:
                    fields = [self._fields[i] for i in indexes]
```

In here `indexes` should not be changed, because they are used later on (L333+) to pick up correct array elements accordingly. 

In your above changes, this test will fail: 

`assert ak._v2.to_list(ak._v2.Array([{"x": [[1, 2, 3], [4, 5, 6]]}])[:, []]) == [{}]` 

Because the slice will output the following:

`ak._v2.to_list(ak._v2.Array([{"x": [[1, 2, 3], [4, 5, 6]]}])[:, []])
[{'x': [[1, 2, 3], [4, 5, 6]]}]`


However, this will mean that when slicing with empty arrays `len(fields)` will be > 0 while `len(contents) == 0` and will cause a `ValueError` at L96 which can be fixed with `if len(contents) > 0 and not len(contents) == len(fields):` (in the issue, @jpivarski pointed out that even if the length is 0, the field should be passed on). 

Moreover, accepting different lengths will require another fix in File "/awkward/_v2/highlevel.py", line 1175, in __repr__ , into
`return self._contents[index] IndexError: list index out of range`

agoose77 replied:
Ah, thanks for this. I've misunderstood the intent of Jim's test case. The empty slice should still drop the contents, but keep the fields (whereas I was leaving the contents for subsequent slicing (that never happens)).

jpivarski replied:
It sounds like this is resolved, right?

(It's complicated and I'm just catching up.)

agoose77 replied:
Yes, it is. I misunderstood the problem at this point.

ioanaif commented on tests/v2/test_1593-empty-slice-list-record.py:
```diff
@@ -0,0 +1,6 @@
+
+def test():
+    array = ak._v2.Array([{"x": [[1, 2, 3], [4, 5, 6]]}])
+    assert array[:, []].fields == ["x"]
```
```suggestion
    assert array[:, []].fields == ["x"]
    assert ak._v2.to_list(ak._v2.Array([{"x": [[1, 2, 3], [4, 5, 6]]}])[:, []]) == [{}]
```

agoose77 committed:
Update tests/v2/test_1593-empty-slice-list-record.py

Co-authored-by: ioanaif <ioana.ifrim@gmail.com>

agoose77 committed:
Fix: convert `[]` into `slice(0, 0)`

This bypasses the `slice_fields` mechanism, which is really orthogonal to the content length `_getitem` implementations.

agoose77 committed:
Refactor: check explicitly for iterables with lengths

agoose77 committed:
Test: fix to match v1

agoose77 committed:
Fix: normalise [] to empty array

agoose77 committed:
Fix: return `NumpyLike.array` instead of index

agoose77 committed:
Fix: support empty Index64

agoose77 committed:
Docs: add small docstring elucidating getitem_broadcast

agoose77 committed:
Fix: update `getitem_next_array_wrap` to accept `outer_length` argument

agoose77 committed:
Fix: remove list special-case

agoose77 committed:
Fix: pass list length to `getitem_next_array_wrap`

agoose77 committed:
feat: raise error if unsupported indexing used

This prevents the second-case of advanced indexing referred to in https://numpy.org/doc/stable/user/basics.indexing.html#basics-indexing:~:text=In%20the%20first,just%20like%20slicing from being used

agoose77 committed:
test: fix use of v1 record

agoose77 committed:
fix: respect zero-length of regular array

agoose77 committed:
test: respect new error during indexing

agoose77 committed:
test: fix expected output of slice

agoose77 committed:
test: fix expected output of slice

agoose77 committed:
test: fix expected output of slice

agoose77 committed:
Merge branch 'main' into agoose77/tentative-fix-1593

agoose77 committed:
fix: return view in `TypeTracerArray.view`

agoose77 committed:
feat: add `TypeTracerArray.astype`

jpivarski commented on src/awkward/_util.py:
```diff
@@ -37,6 +39,10 @@ def is_file_path(x):
         return False
 
 
+def is_sized_iterable(obj):
+    return isinstance(obj, Iterable) and isinstance(obj, Sized)
```
Correct: too often, I assume that iterables have a `__len__` and common test examples don't check it. We're talking about `Sized` `Iterables` in anything we'd support in a slice tuple.

agoose77 replied:
It's a slight pain that there's not an ABC for sized iterables, but it's not too hard to write our own helper, so I should not complain :P

jpivarski commented on src/awkward/_v2/_slicing.py:
```diff
@@ -474,17 +482,44 @@ def prepare_tuple_bool_to_int(item):
+
+
+def index_is_advanced_separator(item):
+    return (item is np.newaxis) or (item is Ellipsis) or isinstance(item, slice)
```
Thinking about the name of this: `np.newaxis`, `Ellipsis`, and `slice` are three examples of things that NumPy would not call "advanced slices." I'm pretty sure the word is only used for arrays in a slice.

Maybe this should be named `is_nonadvanced_separator`?

agoose77 replied:
Haha, I chose the name as these 'primitives' separate advanced array indices! Maybe that is less clear than your suggestion.

jpivarski commented on src/awkward/_v2/_slicing.py:
```diff
@@ -474,17 +482,44 @@ def prepare_tuple_bool_to_int(item):
+    return (item is np.newaxis) or (item is Ellipsis) or isinstance(item, slice)
+
+
+def ensure_supported_tuple(where):
```
This function does check for whether we support a tuple-slice, but it's not the only such place. This:

https://github.com/scikit-hep/awkward/blob/2b1f32404a13e91085db5752b78374a4c380935c/src/awkward/_v2/_slicing.py#L45-L50

ensures that the tuple is supported in the sense of not combining Awkward-indexing (ragged or missing values in arrays) with NumPy advanced indexing (more than one array/int in the tuple, including the Awkward one).

Your function guards against another unsupported case, interleaved advanced and basic NumPy indexing, which is described in the first bullet point and second example of [this section in the NumPy docs](https://numpy.org/doc/stable/user/basics.indexing.html#combining-advanced-and-basic-indexing).

So "`exclude_interleaved_basic_advanced_numpy_indexing`"? It's a mouthful, but the function is only called once.

And maybe the above should be refactored into a function beside this one called "`exclude_mixed_awkward_numpy_advanced_indexing`", so that it's clear that they're both excluding possible cases?

jpivarski commented on tests/v2/test_0397-arrays-as-constants-in-numba.py:
```diff
@@ -100,7 +100,7 @@ def f4():
 
 def test_Record():
-    record = ak.Record({"x": 1, "y": [1, 2, 3]})
+    record = ak._v2.Record({"x": 1, "y": [1, 2, 3]})
```
Yikes! Good catch!

jpivarski commented on tests/v2/test_0024-use-regular-array.py:
```diff
@@ -20,7 +20,7 @@ def test_empty_array_slice():
         == a[2, 1, np.array([], dtype=int)].form
     )
-    assert to_list(a[2, np.array([1], dtype=int), np.array([], dtype=int)]) == [[]]
+    assert to_list(a[2, np.array([1], dtype=int), np.array([], dtype=int)]) == []
```
Your change to this test is correct. Looking at NumPy (and adjusting the definition of `a` so that it's regular),

```python
>>> a = np.array([[[], [], []], [[], [], []], [[], [], []]])
>>> a[2, np.array([1]), np.array([], dtype=int)].shape
(0,)
```

Selecting the third of three `[[], [], []]`'s, the second item within it (`[]`), and then nothing within that (which is the only thing you can do) does return an empty list, not a list of an empty list. That is,

```python
>>> a[2, np.array([1]), np.array([], dtype=int)].tolist()
[]
```

So yes, this one _should_ be changed.

jpivarski commented on tests/v2/test_1568-fix-lengths-empty-regular-slices.py:
```diff
@@ -54,18 +54,17 @@ def test_lengths_empty_regular_slices():
+        _ = e[1:4, [], -3:, []]
+
+    with pytest.raises(ValueError):
+        _ = f[1:4, [], -3:, []]
```
This test looks very deliberate, interleaving advanced and basic slices, but since all the other tests pass after you put in a guard against it, this is the only instance in the whole test suite that violates the rule against interleaving.

It evidently worked for this corner case with empty arrays, but there's no guarantee that it will work in general.

I can say that when the whole getitem system was designed, I intentionally gave up on that interleaving case as something that can't be adapted to variable-length arrays. (The slicers are regular, but they're slicing an Awkward Array that might be anything.) It is the only NumPy rule I consciously gave up on, and I really should have documented that as an explicit prohibition as you just did.

When @ioanaif and I added these test cases 10 days ago, I wasn't thinking about that rule, at least not enough to have realized that these cases are not ones that are expected to work.

So yes, we can change this test to say that this is now an error case.

jpivarski commented on tests/v2/test_1568-fix-lengths-empty-regular-slices.py:
```diff
@@ -92,16 +91,6 @@ def test_lengths_empty_regular_slices():
-        d[1:4, [], -3:, []].shape
-        == ak._v2.to_numpy(e[1:4, [], -3:, []]).shape
-        == ak._v2.to_numpy(f[1:4, [], -3:, []]).shape
-    )
 
```
Also interleaving basic and advanced NumPy slices. It's correct to remove them from the test suite.

jpivarski commented on src/awkward/_v2/contents/content.py:
```diff
@@ -522,13 +522,6 @@ def _getitem(self, where):
                 self._nplike,
             )
 
-            if any(isinstance(x, list) and len(x) == 0 for x in items) and any(
-                isinstance(x, slice) for x in items
-            ):
-                attempt = next.maybe_toNumpyArray()
-                if attempt is not None:
-                    next = attempt
```
It's interesting that it works without this: that RegularArrays that could be a single NumpyArray are being sliced the same way as that NumpyArray, though I know that we put in tests that check for RegularArray-vs-NumpyArray in the corner cases we were concerned about, and they're evidently passing.

(One way to say it is that the `RegularArray._getitem_next` is doing what it's supposed to do. We thought that maybe this was a problem for edge cases, but all of those edge cases are in the test suite.)

The tests that have been removed or changed in the test suite are not ones that this is addressing. So it's fine to remove this code, as you have.

jpivarski reviewed (changes requested):
I went over this with a fine-toothed comb because getitem is complicated, and I know you said you were changing some of the expected test results. But (as you can see below), I think everything here is correct.

Some things are messy, like the normalization of `Index64` vs `np.ndarray` (or CuPy, etc.) before the last step of normalization, but the normalization is correct before entering `_getitem_next`, which is the important part.

The whole getitem system can stand to have some developer-focused documentation, or better naming, internal function docstrings, comments, or (if they can be made to do anything at all, given the size of the codebase) type annotations. The getitem system had once been under heavy development, but now it's rather stable and that sort of documentation would not go out of date quickly.

I can also say that it's 1 order of magnitude better than v1's pre-processing of slices and 2 orders of magnitude better than v0's handling of slices. We're converging.

The only actual changes I'm requesting are renaming two functions for clarity (below).

agoose77 committed:
feat: raise `ValueError` for mixed indices

This changes #1305

agoose77 committed:
refactor: move NumPy index validation into broadcasting

This also renames various utility functions to make the concerns more apparent.

agoose77 committed:
fix: actually raise exception

agoose77 committed:
Merge branch 'main' into agoose77/tentative-fix-1593

agoose77 committed:
refactor: remove unneeded length check

agoose77 committed:
Merge remote-tracking branch 'origin/agoose77/tentative-fix-1593' into agoose77/tentative-fix-1593

jpivarski reviewed (approved):
These names and error messages are much better.

Your functional validator is quite nice! I don't have an explicit preference for inline iterators; it's just that that was quicker to write for not having to invent a bunch of new names. (Names are the hardest part.)

(The [Awkward 0 getitem, just for JaggedArray](https://github.com/scikit-hep/awkward-0.x/blob/dd885bef15814f588b58944d2505296df4aaae0e/awkward0/array/jagged.py#L517-L787) gets pretty deep and still can't even handle a lot of cases.)

This is ready to squash-and-merge!
